### ISSUE TITLE
[macos] Unstable loop timing with default loop mode on macos 

### ISSUE THREAD

**Issue by mitchmindtree** [nannou-org/nannou#250]
@JoshuaBatty and I have been trying to track down why he has been getting a low frame rate on his macbook pro in a multi-window downstream commercial project and have been doing some profiling. We timed the main loop and noticed that there was a wide variety from ~9ms to 32ms in interval between `update`s.

To check our sanity, we decided to time the nannou `simple_window` example `update` interval and noticed that we could still see the same unstable loop rate, even though the example renders no graphics apart from clearing the swapchain image. We discovered that by far the largest amount of time is actually spent on the `.then_signal_fence_and_flush()` call where the metal runtime waits on a semaphore as it attempts to acquire the next swapchain image (the `nextDrawable` function in metal). When I run the `simple_window` example on my Linux machine, I get a much more consistent timing with almost exactly 16ms between calls to `update`.

One possibility is that the macos swapchain image simply takes a long time to present, and rather than waiting for this presentation to finish before presenting the next image (as is currently the case in nannou) users are expected to present the next image in the meantime. To account for this, I tried changing the `previous_frame_end` `Option<GpuFuture>` field to a queue of previous frame ends. This allows us to process `swapchain.num_images()` of frames before ever waiting for a previous frame. Once the `previous_frame_ends` queue contains `swapchain.num_images()`, we pop the oldest and wait on that. Unfortunately, this didn't seem to have any effect on the default application loop interval consistency on @JoshuaBatty's MBP, even when we increased the number of images in the swapchain from 2 to 3. Please see my [`swapchain_image_queue`](https://github.com/mitchmindtree/nannou/tree/swapchain_image_queue) branch if you would like to test for yourself.

cc @freesig if add an `update` function to the `simple_window.rs` example and add
```rust
println!("{}", update.since_last.ms());
```
do you notice the same unstable intervals?

**Comment by mitchmindtree** [comment]
Nah no significant recursion anywhere, it's the blocking on the metal `nextDrawable` obj-c function call which gets called inside `.then_signal_fence_and_flush()` (called on the acquire_next_image future) that is taking so long. [Here's the docs for `nextDrawable`](https://developer.apple.com/documentation/quartzcore/cametallayer/1478172-nextdrawable) - I'm guessing the varying time is how long it takes the image to finish presentation, though could be wrong.

**Comment by freesig** [comment]
ok I'm getting between 16-19ms but im on the old sdk. Ill try updating

**Comment by freesig** [comment]
Ok after updating the molten sdk I am now getting ~8ms which is probably my screens refresh rate. I just did a profile as well and I'm not getting the same `.then_signal_fence_and_flush()` block. 
I could be something to do with @JoshuaBatty graphics drivers maybe? Because he has a dedicated card and I just have the integrated

**Comment by mitchmindtree** [comment]
Closing this as the issue seems to have disappeared on joshy's MBP now! It's possible he picked up a new moltenvk version or one of the recent PRs fixed it. We tried rolling back a couple of commits on his machine but couldn't find it with a quick check.

If you're reading this and are having frame-rate consistency issues on your own macos machine, feel free to ping this issue and I'll re-open it!
